### PR TITLE
KAFKA-10539: Convert KStreamImpl joins to new PAPI

### PR DESCRIPTION
**Pull request by jeqo**
As part of the migration to new Processor API, this PR converts KStream to KStream joins.

Depends #11315 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java [hunk 1]
```diff
@@ -200,11 +217,11 @@ private void emitNonJoinedOuterRecords(final KeyValueStore<TimestampedKeyAndJoin
             // throttle the emit frequency to a (configurable) interval;
             // we use processing time to decouple from data properties,
             // as throttling is a non-functional performance optimization
-            if (context.currentSystemTimeMs() < sharedTimeTracker.nextTimeToEmit) {
+            if (internalProcessorContext.currentSystemTimeMs() < sharedTimeTracker.nextTimeToEmit) {
```
Comment (jeqo) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:220]: @vvcephei, I haven't found a way to get the current system time. Is there a better alternative than this?

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:220]: This was an oversight in the kip that added "current system time" to the public API. They only added it to the deprecated ProcessorContext. I planned to send a PR to fix it, but for now this is a fine solution.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java [hunk 2]
```diff
@@ -108,26 +116,35 @@ public void init(final org.apache.kafka.streams.processor.ProcessorContext conte
...
+                if (context().recordMetadata().isPresent()) {
+                    final RecordMetadata recordMetadata = context().recordMetadata().get();
+                    LOG.warn(
+                        "Skipping record due to null key or value. "
+                            + "topic=[{}] partition=[{}] offset=[{}]",
+                        recordMetadata.topic(), recordMetadata.partition(), recordMetadata.offset()
+                    );
+                } else {
+                    LOG.warn(
+                        "Skipping record due to null key. Topic, partition, and offset not known."
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java]:
```suggestion
                        "Skipping record due to null key or value. Topic, partition, and offset not known."
```

**Review by vvcephei — COMMENTED** [review]
Thanks @jeqo ! This looks good to me. There's just one typo in the log message.

**Comment by vvcephei** [comment]
Test failure was unrelated.

**Review by vvcephei — APPROVED** [review]
Thanks, @jeqo !

**Pull request merged**